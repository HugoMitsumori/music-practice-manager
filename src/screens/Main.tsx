import React, { useEffect, useState } from 'react'
import { StatusBar } from 'expo-status-bar'
import { StyleSheet, View } from 'react-native'
import AsyncStorage from '@react-native-async-storage/async-storage'
import {
  Button,
  Text,
  TextInput,
  Provider as PaperProvider,
} from 'react-native-paper'

const LIST_KEY = 'list'

const Main = (): JSX.Element => {
  const [list, setList] = useState([] as string[])
  const [listItem, setListItem] = useState('')

  const addCurrentItem = () => {
    if (listItem) {
      setList((currentList: string[]) => [...currentList, listItem])
      setListItem('')
    }
  }

  const saveList = async (): Promise<void> => {
    try {
      await AsyncStorage.setItem(LIST_KEY, JSON.stringify(list))
    } catch (err) {
      console.log(err)
    }
  }

  const getSavedList = async (): Promise<Array<string>> => {
    let value
    try {
      value = await AsyncStorage.getItem(LIST_KEY)
    } catch (err) {
      console.log(err)
    }

    if (typeof value === 'string') {
      return JSON.parse(value)
    }

    return []
  }

  useEffect(() => {
    const readSaved = async () => {
      const savedList = await getSavedList()
      if (savedList && Array.isArray(savedList) && savedList.length > 0) {
        setList(savedList)
      }
    }
    readSaved()
  }, [])

  useEffect(() => {
    saveList()
  }, [list])

  return (
    <PaperProvider>
      <View style={styles.container}>
        <View style={styles.actionArea}>
          <Text>Add items to the list!</Text>
          <TextInput label="Item" value={listItem} onChangeText={setListItem} />
          <Button mode="contained" icon="plus" onPress={addCurrentItem}>
            Add Item
          </Button>
          <Button mode="contained" color="red" onPress={() => setList([])}>
            Clear List
          </Button>
        </View>
        {list.map((item: string) => (
          <Text key={item} style={styles.listItem}>{item}</Text>
        ))}
        <StatusBar style="auto" />
      </View>
    </PaperProvider>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    marginTop: 50,
  },
  actionArea: {
    width: '80%',
    height: 200,
    justifyContent: 'space-around',
  },
  listItem: {
    marginVertical: 5,
  },
})

export default Main
