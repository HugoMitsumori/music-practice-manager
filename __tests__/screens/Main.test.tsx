import React from 'react'
import { cleanup, render } from '@testing-library/react-native'
import Main from '../../src/screens/Main'

jest.useFakeTimers()

describe('<Main />', () => {
  afterEach(() => cleanup())
  it('renders correctly', () => {
    const tree = render(<Main />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})