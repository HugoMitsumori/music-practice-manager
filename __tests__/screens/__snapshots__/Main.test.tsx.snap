// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> renders correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flex": 1,
        "marginTop": 50,
      }
    }
  >
    <View
      style={
        Object {
          "height": 200,
          "justifyContent": "space-around",
          "width": "80%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "left",
            },
            undefined,
          ]
        }
      >
        Add items to the list!
      </Text>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "rgb(231, 231, 231)",
              "borderTopLeftRadius": 4,
              "borderTopRightRadius": 4,
            },
            Object {},
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.26)",
              "bottom": 0,
              "height": 2,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "transform": Array [
                Object {
                  "scaleY": 0.5,
                },
              ],
            }
          }
        />
        <View
          style={
            Array [
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              Object {
                "minHeight": 64,
              },
            ]
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <Text
              numberOfLines={1}
              onLayout={[Function]}
              style={
                Object {
                  "color": "#6200ee",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "left": 0,
                  "opacity": 0,
                  "paddingLeft": 12,
                  "paddingRight": 12,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 34,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
            >
              Item
            </Text>
            <Text
              numberOfLines={1}
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.54)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "left": 0,
                  "opacity": 0,
                  "paddingLeft": 12,
                  "paddingRight": 12,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 34,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
            >
              Item
            </Text>
          </View>
          <TextInput
            allowFontScaling={true}
            editable={true}
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder=""
            placeholderTextColor="rgba(0, 0, 0, 0.54)"
            rejectResponderTermination={true}
            selectionColor="#6200ee"
            style={
              Array [
                Object {
                  "flexGrow": 1,
                  "margin": 0,
                  "zIndex": 1,
                },
                Object {
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
                Object {
                  "height": 64,
                },
                Object {
                  "paddingBottom": 4,
                  "paddingTop": 24,
                },
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "textAlign": "left",
                  "textAlignVertical": "center",
                },
                false,
                Array [
                  Object {},
                ],
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#6200ee",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 2,
            "minWidth": 64,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0.75,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 1.5,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <View
              style={
                Object {
                  "marginLeft": 12,
                  "marginRight": -4,
                }
              }
            >
              <Text />
            </View>
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    undefined,
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "#ffffff",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Add Item
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "red",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 2,
            "minWidth": 64,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0.75,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 1.5,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    undefined,
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "#ffffff",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              Clear List
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
